### ISSUE TITLE
Fzy improvement: smart-case

### ISSUE THREAD

**Issue by romgrk** [nvim-telescope/telescope.nvim#215]
I was reading your port of [fzy](https://github.com/nvim-lua/telescope.nvim/blob/master/lua/telescope/algos/fzy.lua) and one easy improvement over the original fzy that I always do (yes I've ported it multiple times) is to avoid converting the string to lowercase if the needle contains an uppercase character. It improves the results a lot when the user knows there is an uppercase character in there.

**Comment by tjdevries** [comment]
@swarn you wanna check this out?

**Comment by tjdevries** [comment]
Also, I'm wondering if `vim.stricmp` might be useful for us, not 100% sure.

**Comment by swarn** [comment]
No problem. I'll send a PR in the next few days.

**Comment by swarn** [comment]
And I see that @romgrk is adding lua bindings to the `fzy` C libs that are an order of magnitude faster...

**Comment by romgrk** [comment]
Btw you're going to run just so much faster if you use native. It's really trivial with luajit's ffi, I've done an adaptation here: https://github.com/romgrk/fzy-lua

It contains your original port of fzy (original.lua) and ffi bindings to fzy (native.lua). The repo includes the source for fzy, you just need to `make` to build the dynamic library. For comparison, you can run `luajit benchmark.lua`, on my machine I get the following results:

```
Lines: 100000

Native:
 -> Total: 34130
 -> Time:  69.418 ms
Original:
 -> Total: 34130
 -> Time:  835.683 ms
```

More than an order of magnitude gain.

Currently `native.lua` is a 1:1 map to your `original.lua`, and I've added to both file an `fzy.filter(needle, haystacks)` function to run my benchmarks.

I've also spotted one misusage of FZY in your port. Your `fzy.positions` function does not return the score. I've understood `lua/telescope/sorters.lua` correctly, you're calling `fzy.score` and `fzy.positions` separately. That's no good for performance because both functions call the internal fzy `compute` function, which does the exact same thing in both cases. So you're running `compute` twice by not returning the score in `fzy.positions`, as the original fzy does.

Note that C version of fzy in that repo contains 2 differences with the original:
 - the functions take a `is_case_sensitive` additional parameters that does what you think. It's a parameter rather than computed in the function because for our use-case, you want to compute `is_case_sensitive` once for needle, instead of on each iteration of the loop (because we filter lots of things).
 - `match_positions` (that you call `fzy.positions`) uses `uint32_t` for positions, because `size_t` (`uint64_t`) doesn't map well to lua types

I'm not sure I want to do a PR for this because I myself use clap (for which my bindings will surely be re-used) but I'd be open to assist with adapting the repo to your use.

**Comment by swarn** [comment]
Since you've worked with `fzy` before @romgrk, maybe you can confirm for me: since the C implementation is based on a (very fast) [lookup table](https://github.com/jhawthorn/fzy/blob/master/src/bonus.h), it only works for ASCII, yes?

**Comment by romgrk** [comment]
Didn't thought about that but yes, that would make sense. Fzy's latest release (in 2018) says they support utf8 though, and I forked their code recently. It would probably be necessary to test out with non-ascii inputs.

If you guys are interested we could work together on that, as I said it's probably going to be useful for clap as well so I have some interest in this.

**Comment by romgrk** [comment]
Update: the implementation doesn't choke on utf8 more-than-one-byte character, but it does consider those bytes as single characters and returns, eg 2 positions if those bytes match.

```lua
local native = require'native'

local positions = native.positions('pé', 'spécial')

for i, p in ipairs(positions) do
  print(p)
end

-- Outputs: (zero-indexed)
--  1 (p)
--  2 (é first byte)
--  3 (é second byte)
```

That seems similar to lua's `string:sub` that will substring by byte (from what I can see) rather than by character.

**Comment by tjdevries** [comment]
@swarn I may add a new way for people to opt-in to compiled native code.

I will probably add @romgrk code which will load if compiled, otherwise will use the pure lua sorter. This way we can allow people who do not want to compile / build anything to continue doing so, but for people who want to go 10x faster, they can do that easily :)

**Comment by romgrk** [comment]
Yes, that sounds good. What do you think we work on that on my repo, and make it a dependency of telescope/clap? It would be nice to expose it as a self-contained plugin that would export the native functions if they're available and fall back to lua. It could also expose `.is_installed()`/`.install()` to facilitate usage, and the build process should be improved to support macOS & Windows.

**Comment by swarn** [comment]
This issue's scope has grown quickly! I _was_ just going to add a few checks for uppercase chars inside the `fzy` functions, but now that seems sub-optimal:

1. As @romgrk points out, you want to check the `prompt` for uppercase chars only once per `prompt`, rather than once per `line`.

2. Right now, I think that the picker calls `sorter.scoring_function(prompt, line)` and `sorter.highlighter(prompt, line)` once per entry, yes? That supports progressively building the sorted match list as new entries asynchronously come in.

3. That means the check for uppercase in `prompt` would need to occur in the picker, I think. And then, the `sorter` interface functions would need (optional) `is_case_sensitive` parameters.

While I'm mentioning changes to the `sorter` interface, a `sorter.score_and_highlight(...)` function would allow skipping the redundant computation in `fzy` scoring and highlighting, without requiring caching inside the sorter.

I've got limited time at the moment, and @romgrk is moving quickly, so don't wait on me.

**Comment by tjdevries** [comment]
Here's where I'm thinking about it at the moment : https://github.com/nvim-lua/telescope.nvim/pull/217

I'm not opposed to working on it in another plugin, but I also may add other ffi based options for telescope, so it wouldn't be completely wasted to manage this here for me. 

I don't have much of a preference about it, tbh. I may modify the code further, but I do not know at the moment.

**Comment by tjdevries** [comment]
I was considering experimenting with Lua rocks style compilation to add it that way.

**Comment by romgrk** [comment]
Not sure how luarocks compilation works, it could be a good option. On my side I was thinking about bundling [prebuilt binaries](https://github.com/romgrk/fzy-lua-native/tree/master/static) for the major platforms (included windows x64 & linux x64 already). I've set it to to be differentiated by os & architecture, it might also need to be differentiated by libc but I don't have enough experience at this point to tell. This has the advantage of skipping the compilation step for hopefully most users, and the ones with more exotic systems usually have compilation tools available anyway. I don't have acces to a mac but I was thinking maybe building the missing platforms through travis. 

I've also completed the binding if you want to use `match` instead of `match_positions` for `fzy.score`. Also [this](https://github.com/nvim-lua/telescope.nvim/pull/217/files#diff-ad54f7a0c392aafe66950a4b05344998019c3ebd894a526a2e65cddc392d63f1R183-R192) and [this](https://github.com/nvim-lua/telescope.nvim/pull/217/files#diff-ad54f7a0c392aafe66950a4b05344998019c3ebd894a526a2e65cddc392d63f1R133-R139) should probably be uncommented.

IMHO joining efforts to fine-tune the fzy repo would be ideal because it's less lost effort, but I get how it's faster to just include the source yourself so I won't hold any grudges 😊

**Comment by tjdevries** [comment]
Ah, cool idea & thought w/ the pre-built binaries stuff.

Now I'm kind of wondering if we should make some repo / library for this kind of thing generally. A bunch of the work could be generalized to a skeleton for (particularly low-dependency) C compilation could be done and we could write some helper functions for loading them if they exist.

In addition, you can use luarocks to declare different modules you have within one repository! So we could do something like `src/fzy-native`, `src/other-thing`, etc. within the same repository. We could publish to luarocks (for general use case if people were interested), make the prebuilt static binaries (perhaps as releases? using gh actions or something similar?), etc. all within one shared library.

What do you think about the general idea?

**Comment by romgrk** [comment]
I like the idea. Sounds like [node-pre-gyp](https://github.com/mapbox/node-pre-gyp), which is a bit hard to setup but is very similar (using travis + S3 buckets). But github actions and releases would be an interesting target. I'm in. How do you want to proceed? I would try to set up my fzy repo as a template because it's minimal enough, and from that extract an helper module. Let me know if you have another idea.

**Comment by tjdevries** [comment]
Let me play with the idea and Lua rocks. Might do it on stream tonight or tomorrow and then I'll check back in.

**Comment by swarn** [comment]
I've updated my [pure Lua](https://github.com/swarn/fzy-lua) implementation; it sounds like I should hold off on anything else until other decisions are made.

It makes sense to have `fzy` structured so that it's available for automatic consumption as a dependency, in either form. My code is structured the way it is purely because I didn't know any better: this is the only thing I've ever written in Lua.

Does it make sense to have the C and Lua versions in the same repo? I'm not really looking to be an upstream dependency maintainer, I'm happy to put `fzy-lua` up for adoption.

**Comment by swarn** [comment]
I've got the pure Lua version up [on LuaRocks](https://luarocks.org/modules/sethwarn/fzy); I'm currently figuring out how to get the luajit version on there, and allow it to automatically use the compiled lib and fall back on the lua version.

**Comment by romgrk** [comment]
I've already written some logic to load the compiled lib [here](https://github.com/romgrk/fzy-lua-native/blob/master/lua/native.lua#L10-L24). We should also ensure it's loadable through both neovim and regular luajit, which use different defaults for `package.path`.

**Comment by romgrk** [comment]
I've completed the github action [workflow configuration](https://github.com/romgrk/fzy-lua-native/blob/master/.github/workflows/main.yml) to use it to build on ubuntu, macos & windows. The binaries are [there](https://github.com/romgrk/fzy-lua-native/tree/master/static) but I have only tested on (arch)linux for now. I can spin a windows VM to test it but I don't have access to a macos env. If any of you uses windows or macos as their env you could help testing just by running `luajit benchmark.lua` with the package. Now that I think I should have set the GH workflow to do it. I might add that test.

I was not sure how to expose the package, I've put the files in a `lua` folder, but I think the default search path is something like `package-name/init.lua`. I'm also not sure how to set it up in relation to fzy-lua. Should fzy-lua-native try itself and fallback on fzy-lua, or should fzy-lua do it? For the moment fzy-lua-native includes a copy of @swarn's port.

**Comment by swarn** [comment]
@romgrk, the github actions are slick!

I tried it on my mac, got an error, but I might not be running it correctly:

    $ git clone https://github.com/romgrk/fzy-lua-native.git
    $ cd fzy-lua-native
    $ make
    $ luajit benchmark.lua

    luajit: ./lua/native.lua:24: dlopen(./lua/../static/libfzy-osx-x86_64.so, 5): image not found
    stack traceback:
        [C]: in function 'load'
        ./lua/native.lua:24: in main chunk
        [C]: in function 'require'
        benchmark.lua:9: in main chunk
        [C]: at 0x010751ca2c

In the meantime, in a possibly wasted duplication of effort, I've updated my repo with a native version. It uses the regular lua C interface, rather than the vastly cooler luajit ffi, but that does mean it runs under both, and luarocks seem to handle building everything.

**Comment by swarn** [comment]
@romgrk Ah, it's looking for `libfzy-osx-x86_64.so`, but it generated `libfzy-darwin-x86_64.so`.

**Comment by romgrk** [comment]
@swarn I've added the translation for osx => darwin just now, it should work now. The lua C interface is also a nice option. I thought about it before doing the binding but I targeted luajit's ffi as it's probably easier to do it there and neovim includes luajit on all platforms. The lua C interface is probably more adapted for a package on luarocks though.

I don't really have an opinion on which is better between the C interface or luajit's ffi, both are probably fine. If you want to reuse the github workflow configuration go ahead, it took a bit of time to configure because many GH actions were written for v1, and only v2 actions work consistently across all OSes, so I had to port [this action](https://github.com/romgrk/publish-to-github) for usage in the config.

**Comment by swarn** [comment]
Yeah, it works perfectly now, nice 👍 

I like the current setup for `fzy-lua`: I don't have to publish any binaries, `luarocks` builds them on the client machines. I have no idea how that interoperates with the neovim lua ecosystem, though.

I do need to automate some testing. My test suite showed me something I didn't realize: default `fzy` is half-case-sensitive, so lower-case matches to upper-case, but not vice-versa. This is not the same behavior as his javascript implementation or my lua implementation. 

For `fzy-lua`, I changed the C implementation (stolen from you) to match the lua implemenation: either fully case-sensitive, or not at all.

Also, re: exposing the package, I built them as `fzy_lua` and `fzy_native` modules, with a `fzy` module that tries native before falling back to lua.

**Comment by romgrk** [comment]
One suggestion in case your package is used, one improvement attempt I made in [node-fzy](https://github.com/romgrk/node-fzy) was to expose functions such as `match_positions(needle: String, haystacks: Array<String>)`. It could even go to exposing a function such as `filter(needle: String, haystacks: Array<String>)` to apply both `has_match` and `match_positions` on a list of strings. The idea was to avoid switching back and forth between JS and C. From my testing in javascript it did not improve benchmarks substantially, and in some cases it the performance even dropped. I don't know enough about the lua/luajit VMs to say what would be the effect of that but it would be nice to try it.

> For fzy-lua, I changed the C implementation (stolen from you) to match the lua implemenation: either fully case-sensitive, or not at all.

It is intentionally half-case sensitive. You're talking about smart-case right? Because if the users types `myFile`, there is a clear intent to search for a file with a capital F in it, because it's more work to do so, but this behavior improves the quality of matches. (also: the title of this thread ^^)

**Comment by swarn** [comment]
Luarocks doesn't _require_ a compiler? Seriously, I had to go look. The first thing in their linux install instructions is `apt install build-essentials`, but the homebrew formula for luarocks doesn't require one. I will definitely steal your github action and generate some binaries.

Regarding case-sensitivity, I did notice the title of the thread 😄. And the one-way sensitivity behavior was clearly Hawthorne's intention — though as I said, he ditched it in his js implementation.

However, it doesn't match vim's `smartcase` behavior. In vim, I think, once you add any capital letters to a search, the whole thing becomes fully case-sensitive. Searching `aaa` will find `AaA`, and searching `Aaa` will not, but original flavor `fzy` will match. I can't imagine it makes much practical difference either way, but I like matching vim 🤷. 

You're 100% correct about adding a vector version of the functions, e.g. `matches(needle, bunch_o_haystacks)`. I'll have to try some testing...

**Comment by romgrk** [comment]
Seems to make a difference for luajit:

![Screenshot from 2020-11-05 01-35-56](https://user-images.githubusercontent.com/1423607/98206727-68009480-1f08-11eb-9e56-a2383e7d6091.png)

**Comment by tjdevries** [comment]
OK, so I'm still not 100% sure how I'd like to do this.

One thing is perhaps something can be made in the new org for https://github.com/nvim-telescope/ where we can implement some optional sorters and point people to them. I'm picturing that we have the pure lua version in this repo (or maybe move it? I'm not sure -- it's nice to have a few decent builtins), but the pure lua version checks if something (say, https://github.com/nvim-telescope/telescope-fzy-native.nvim) is installed and uses that instead if so.

users can install the other plugin if they want and have their plugin manager run `make` after they install to do the compilation (or luarocks build, or whatever it takes to make that happen).

What do both of you think?

**Comment by romgrk** [comment]
For clap I was leaning towards distributing the fzy module directly with it, either as a git submodule or just as a vendored copy of it. The prebuilt binaries for all 3 major OSes are there (but untested except for linux), so for most users there shouldn't even need to be a build step required. The only thing missing in my repo would be the logic for loading the non-native implementation if the native one fails.

For telescope I don't have opinions :] Do whatever works for you.

**Comment by swarn** [comment]
What's the status of using luarocks as a dependency? I see work like neorocks and packer.nvim; is the neovim community moving towards assuming luarocks are available (like how coc plugins rely on npm)?

If so, I realize now that I was exactly wrong a few weeks back — I _would rather_ maintain `fzy-lua` in its own repo, with its own tests, etc. It's building on all platforms and (irrelevant to nvim, I know) all luas. I've made a few tweaks, and used romgrk's excellent suggestions. It does the check for a working native version and falls back to the pure lua version if needed. So I think it's pretty stable, and it seems easier for you to remove all of that from `telescope`.

If not, that's tricker. I like the "check for other, native plugin" idea. Would `telescope-fzy-native` just download prebuilt binaries?

**Comment by romgrk** [comment]
> What's the status of using luarocks as a dependency? 

Dependency management is hell in neovim, just as vim -_- It's way easier to just include your dependecies directly

> Would telescope-fzy-native just download prebuilt binaries?

I'd advise against adding build/download steps, knowing that the size of the binaries is about 50KB each, way easier for your users to just include everything upfront

**Comment by swarn** [comment]
I see that `vim-clap` defaults to vimscript, provides `:Clap install-binary` to use `cargo` to build binaries, **and** `:Clap install-binary!` to try precompiled binaries, which is slick.

**Comment by clason** [comment]
> Dependency management is hell in neovim, just as vim -_- It's way easier to just include your dependecies directly

Yes, but people are actively working to make that better. (Don't forget that vim builds on a fantastic billion of platforms. Nvim cuts down on that, but you still don't want to keep track.)

Indeed, the goal is to have luarocks dependencies in packer (although @romgrk is correct; the binaries are so small that it doesn't make sense to have literal bits off install size by trying of a shared system library; I'd assume every plugin will have independent dependencies (ha!)).

Until (and even) then, if you go for a `telescope-fzy` plugin, I think you're free to handle this however you wish (including bundling binaries, like `fzf` does). My personal view is that for binaries, each platform knows best how to install binaries, so requiring users to make sure it can be found in `$PATH` (however you would do that on a specific platform) is perfectly fine.

**Comment by tjdevries** [comment]
For now, I added the extension that will allow using the fzy-native from romgrk at the moment because it ships with the binaries already done (which makes it real easy to use for me as an extension). So I'm going to close this issue. We can re-open if we need with more stuff.
